### PR TITLE
Setup associations

### PR DESCRIPTION
**Pull request by SimbaWANGU**
In this Pull Request, the following changes were implemented:
1. Created a class Classroom with the following:
- @label instance variable, initialized in the constructor.
- Setter and getter for @label

2. Created the has-many/belongs-to relationship between Classroom and Student. The following were implemented:
- Created the has-many side (a classroom has many students).
- Created the belongs-to side (a student belongs to a classroom).
- Made sure that when adding a student to a classroom it also sets the classroom for the student.
- Made sure that when setting the classroom for a student it also adds it to the classrooms' students.

3. Created a class Book with the following:
- @title and @author instance variables, initialized in the constructor.
- Setters and getters for instance variables

4. Created a class Rental with the following:
- @date instance variable, initialized in the constructor.
- Setter and getter for @date

5. Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following were implemented:
- Created the has-many side of Book and Rental (a book has many rentals).
- Created the belongs-to side of Rental and Book (a rental belongs to a book).
- Created the has-many side of Person and Rental (a person has many rentals).
- Created the belongs-to side of Rental and Person (a rental belongs to a person).
- Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals.push(rental)
+    rental.book = self
```
Comment (FrederickMih) [inline book.rb:12]: - [ ] Let's create an instance of the Book class inside this method and sent the book itself as a parameter in order to achieve the has-many association. 
**Hint:** `Rental.new(person, date, self)`
I think it's better to do it like this.
- [ReadMore...](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)

File: rental.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class Rental
+  attr_accessor :date, :book, :person
+
+  def initialize(date, book, person)
```
Comment (FrederickMih) [inline rental.rb:4]: - Great!. You have achieved the many-to-many relationship by setting in Book and Person in the Rental constructor :+1:

File: person.rb [hunk 3]
```diff
@@ -10,6 +10,7 @@ def initialize(age, name = 'Unknown', parent_permission: true)
     @name = name
     @age = age
     @parent_permission = parent_permission
+    @rentals = []
```
Comment (FrederickMih) [inline person.rb:13]: - A person has many rentals. This is fantastic :+1: 

- [ ] Let's now make sure that there's a method for adding rental in this Person class. [ReadMore...](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)

- In this method, let's kindly make sure that when creating a new Rental, a person itself should be sent as a parameter to achieve the has-many association. [ReadMore...](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)


 









Good L

**Review by FrederickMih — CHANGES_REQUESTED** [review]
Hi @SimbaWANGU , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



# To Highlight, 

-  GitHub workflow has been followed :heavy_check_mark: 
-  PR has a title and summary :+1: 
-  Linters checks are OK :heavy_check_mark: 

## Required Changes ♻️

- [ ] Let's kindly fix the License link in the README.md  file. The link there right now is broken. Fixing it will render your README file a professional one.

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classroom.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class Classroom
+  attr_accessor :label, :students
```
[the remark this answers is not in the document]
Comment (SimbaWANGU) [reply classroom.rb:2]: Done

File: student.rb [hunk 5]
```diff
@@ -1,12 +1,23 @@
 require_relative 'person'
 
 class Student < Person
-  def initialize(classroom, age, name = 'Unknown', parent_permission: true)
-    super(age, name, parent_permission)
+  attr_accessor :classroom, :name, :age
```
[the remark this answers is not in the document]
Comment (SimbaWANGU) [reply student.rb:4]: @FrederickMih If I set that to read only then how will I go about updating the student classroom? In the [study material](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md) the :owner uses `attr_accessor` in the animal class.

image

File: book.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+class Book
+  attr_accessor :title, :author, :rentals
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
```
[the remark this answers is not in the document]
Comment (SimbaWANGU) [reply book.rb:10]: Done

File: book.rb (hunk 1 above)
Comment (SimbaWANGU) [reply book.rb:12]: Done

File: person.rb (hunk 3 above)
Comment (SimbaWANGU) [reply person.rb:13]: @FrederickMih I'm not sure what you're asking me to do here. I have followed the study material and read through the project requirements and these don't sound like they are part of the project requirements. In addition the links you have attached are directing me to the one to many relationships where as I was implementing many to many relationships which is described in the next section of the same link.

**Comment by SimbaWANGU** [comment]
@danmainah, thank you for the clarification. 👍

**Review by mikethreels — APPROVED** [review]
Hi Simba, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**